driver/bpf/filler_helpers.h

```diff
@@ -873,7 +873,7 @@ static __always_inline int __bpf_val_to_ring(struct filler_data *data,
 	case PT_CHARBUF:
 	case PT_FSPATH:
 	case PT_FSRELPATH: {
-		if (!data->curarg_already_on_frame) 
+		if (!data->curarg_already_on_frame && val)
```

Comment:
@FedeDP @Andreagit97 Please let me know if that is fine here.

Reply:
Makes sense imho.

Reply:
Not sure about that, if `val==NULL` we set `len = val_len` but val_len could be something different from `0`, it shouldn't but we need to check for all its usage otherwise the semantics would be different from today. With the actual implementation, we are sure that if `val==NULL`, `len` would be `0` because we have this :point_down:  `else` case, WDYT?
```
if(res >= 0)
{
     len = res;
}
else
{
     /* This should be already `0`, but just to be future-proof. */
     len = 0;
}
```

Reply:
Oh actually, andrea is right!

Reply:
with this new implementation, we rely on the `__bpf_val_to_ring` consumer, it should always put `val_len=0` when `data->curarg_already_on_frame=false`... I would probably keep the old logic since in this case we always have the enforcement unless you spot some issues with the actual implementation :thinking:

Reply:
> With the actual implementation, we are sure that if `val==NULL`, `len` would be `0` because we have this point_down `else` case, WDYT?

For `val==NULL` (or any other  value) and `!data->curarg_already_on_frame`, the `bpf_probe_read_kernel_str` is always performed. And according to the tests, I receive a `res == 1` which then would result in `val==0` and `len==1`. This then resulted in an invalid parameter length.

I agree on the changed semantics here. So my proposal here would be to only perform the `bpf_probe_read_kernel_str` if `val` is non-zero, i.e., pull the check into the block itself.

Reply:
> I receive a res == 1 which then would result in val==0 and len==1. This then resulted in an invalid parameter length.

Uhm I didn't understand this, why `len == 1`? we should obtain `len == 0` if `val==NULL`.

Reply:
> Uhm I didn't understand this, why `len == 1`? we should obtain `len == 0` if `val==NULL`.

Unfortunately, not on `s390` architecture :wink: 

I guess the assumption is that reading from address zero causes a page fault etc. so that zero is returned -- I assume that is the default on x86, arm, other architectures.

On `s390` arch, there is the concept of what we call `low-core` (or `prefix area`). It is an area of 2 pages with lots of critical state information for system operation. Each CPUs has this information at the same place and it needs to be set up with `prefixing` / `set prefix` instructions.

Reply:
Supercool that is what I want to hear :joy: 
So just to summarize our current behavior

* x86 (modern bpf and bpf)
  - NULL POINTER (const char* pointer = NULL;)
   `bpf_probe_read_user_str` returns EFAULT so we return an empty param to userspace
  - EMPTY STRING (const char* pointer = " ";)
   `bpf_probe_read_user_str` returns 0 so we return an empty param to userspace

* x86 (kmod)
  - NULL POINTER (const char* pointer = NULL;)
   `ppm_strncpy_from_user` returns `-1` so we return an empty param to userspace
  - EMPTY STRING (const char* pointer = " ";)
   `ppm_strncpy_from_user` returns 1 so we send an empty string to userspace "\0"

* s390x (modern bpf and bpf)
  - NULL POINTER (const char* pointer = NULL;)
   `bpf_probe_read_user_str` returns `1`  so we send an empty string to userspace ("\0")
  - EMPTY STRING (const char* pointer = " ";)
   `bpf_probe_read_user_str` returns 0 so we return an empty param to userspace

* s390x (kmod)
  - NULL POINTER (const char* pointer = NULL;)
   `ppm_strncpy_from_user` returns `1`  so we send an empty string to userspace ("\0")
  - EMPTY STRING (const char* pointer = " ";)
   `ppm_strncpy_from_user` returns 1 so we send an empty string to userspace "\0"

* ARM64 (I need to check but IIRC we have the same behavior of x86)

is this right?

Reply:
hmm... a `bpf_probe_read_user_str` depends on whether user space maps something at address zero. The `prefix area` is kernel memory. For an empty string `""` I would assume that all return a `\0`  (depending on the memory where the empty string (`\ü`) is.

Reply:
> For an empty string "" I would assume that all return a \0 (depending on the memory where the empty string (\0) is.

I would expect the same thing but this doesn't seem to be true, at least on x86 :(

Reply:
> hmm... a bpf_probe_read_user_str depends on whether user space maps something at address zero.

yeah sorry I wrote `bpf_probe_read_user_str` instead of `bpf_probe_read_kernel_str` :facepalm: 
BTW let's address the `NULL` case first, now I understood your initial change, I didn't know the concept of `prefix-area` on s390x. In this case, yeah I would suggest putting the check on `val` inside the block and enforcing `len=0` :)

Reply:
Just updated the commit with that in mind. Yeah, `prefixing` is very s390 specific... let me know if you like to have learn more in that area ;-)

Reply:
Thank you!
Just to recap for kernel memory if we have a NULL pointer now we should be compliant between all our 3 drivers both for `CHARBUF` like cases both for `BYTEBUF`.
* modern bpf already addresses the null pointer for both cases
* kmod already addresses the null pointer for both cases
* bpf already addresses the null pointer for BYTEBUF case, and after this PR it will address also the CHARBUF one

Reply:
I've have the issue here https://github.com/falcosecurity/libs/issues/866